### DISCUSSION TITLE
How to generate Faust code

### DISCUSSION THREAD

**Discussion by jaapnoordzij** [pschatzmann/arduino-audio-tools#1565] [Q&A] [answered]
Hi Phil
I am in the process of trying to get some Faust code working but so far all my test have failed most of the time compiling OK but crashing during execution. In order to eliminate errors from my side as much as possible I have first run one of your examples (streams-faust_flute-i2s.ino). It works as expected.

After that I have left the sketch unchanged but re-generated the Faust code as instructed in the Wiki :
`>faust -mem fluteMIDI.dsp -o fluteMIDI.h`

The sketch compiles OK but crashes with:
```
[E] AudioFaustDSP.h : 230 - allocate 262144 bytes - failed
Guru Meditation Error: Core  1 panic'ed (StoreProhibited). Exception was unhandled.

Core  1 register dump:
```
etc...

I am running Faust on my Mac (Sonoma).
I have installed it with `brew install Faust`

The header of the generated fluteMidi.h starts with:
```
/* ------------------------------------------------------------
copyright: "(c)Romain Michon, CCRMA (Stanford University), GRAME"
license: "MIT"
name: "FluteMIDI"
Code generated with Faust 2.72.14 (https://faust.grame.fr)
Compilation options: -lang cpp -mem -ct 1 -es 1 -mcd 16 -mdd 1024 -mdy 33 -single -ftz 0
------------------------------------------------------------ */
```

Do you know what is wrong here ?
Kind regards, 
Jaap Noordzij

**Comment by pschatzmann** [comment, marked as answer]
Microcontrollers have only a very small amount of RAM and usually you cant allocate big blocks of memory.
Fortunately some ESP32 have PSRAM. So, if you one with PSRAM you need to activate it in Arduino Tools menu and in the sketch by activating the memory manager like it is demonstrated in one of the examples by using a #define USE_MEMORY_MANAGER before the includes


Dont forget to add the [-mem option as described in the Wiki](https://www.pschatzmann.ch/home/2022/05/30/esp32-and-psram/)

**Reply by jaapnoordzij** [reply]
Thanks, you are right as always. I did all this but when preparing this test the Arduinoe IDE switched back to some board without PSRAM and I did not notice ...

**Reply by jaapnoordzij** [reply]
When experimenting with Faust I can produce all kinds of nice effects. However the pysical models in the pm library generally produce sound that does not very much sound like intended as compared to running the dsp file through FaustLive.

For example your unmodified example `streams-faust_flute-i2s.ino` does not sound very much like a flute on the AudioKit. It does when I try the dsp file from the example in Faust Live on my Mac.

I have experimented with changing the sample rate, the buffer size, removing the Logger completely etc.
Do you have any suggestions what to try to find a solution or maybe the pm's are simply too complex for the ESP32 ?

I am using the Arduino IDE
PSRAM is used
the logging output looks OK (if turned on, sounds are even worse)
...

**Comment by pschatzmann** [comment]
I guess this is due to the slow speed of PSRAM and the floating point operations.
The only advice would be to keep the sample rate as low as possible
